File: Top-Interview-Questions/Trees-Easy-Collection/102-Binary-Tree-Level-Order-Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// Breadth First Search
// Time Complexity: O(n) has to visit ALL tree nodes
// Space Complexity: O(w) max width of the tree n-1 in the worst case
// Runtime: 8 ms, faster than 49.16%
// Memory Usage: 12.5 MB, less than 62.15%
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
    
        vector<vector<int>> res;
        vector<int> level;
        
        if(!root)
            return res;
            
        queue<TreeNode*> q;
        int nodeCount = 0;
        
        q.push(root);
        
        while(!q.empty()){
            
            nodeCount = q.size();
            
            while(nodeCount > 0){
                TreeNode *node = q.front();
                q.pop();
                
                level.push_back(node->val);
                
                if(node->left)
                    q.push(node->left);
                if(node->right)
                    q.push(node->right);
                
                nodeCount--;
            }
            
            res.push_back(level);
            level.clear();
        }
        
        return res;
    }
};

File: Top-Interview-Questions/Trees-Easy-Collection/102-binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # Sol 1. DFS, Stack
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        result=[]
        if not root:
            return result
        self.dfs(root,0,result)
        
        return result
    
    def dfs(self,root,level,res):
        if not root:
            return 
        if len(res) < level+1: # current level
            res.append([])
        res[level].append(root.val)
        self.dfs(root.left, level+1, res)
        self.dfs(root.right, level+1, res)
    # TC: O(N), n is the number of nodes in the tree
    # SC: O(N), same
    # Runtime: 36 ms, faster than 58.05%
    # Memory Usage: 14.4 MB, less than 16.06%


    # Definition for a binary tree node.
    # class TreeNode:
    #     def __init__(self, val=0, left=None, right=None):
    #         self.val = val
    #         self.left = left
    #         self.right = right
    class Solution:
        def levelOrder(self, root: TreeNode) -> List[List[int]]:
            result=[]
            if not root:
                return result
            self.dfs(root,0,result)
            
            return result
        
        def dfs(self,root,level,res):
            if not root:
                return 
            if len(res) < level+1: # current level
                res.append([])
            self.dfs(root.left, level+1, res)
            self.dfs(root.right, level+1, res)
            res[level].append(root.val)


    # Sol 2. BFS, Queue
